contracts/dutchX/KyberDutchXReserve.sol

```diff
@@ -0,0 +1,349 @@
+    function withdraw(uint amount) public;
+    function deposit() public payable;
+}
+
+
+interface DutchXExchange {
+    // Two functions below are in fact: mapping (address => mapping (address => uint)) public sellVolumesCurrent;
+    // Token => Token => amount
+    function buyVolumes(address sellToken, address buyToken) public view returns (uint);
+    function sellVolumesCurrent(address sellToken, address buyToken) public view returns (uint);
+    function deposit(address tokenAddress,uint amount) public returns(uint);
+    function postBuyOrder(address sellToken,address buyToken,uint auctionIndex,uint amount) public returns (uint);
+
+    function claimBuyerFunds(address sellToken, address buyToken, address user, uint auctionIndex) public
+        returns(uint returned, uint frtsIssued);
+
+    function withdraw(address tokenAddress,uint amount) public returns (uint);
+    function getAuctionIndex(address sellToken, address buyToken) public view returns(uint index);
+    function getFeeRatio(address user) public view returns (uint num, uint den); // feeRatio < 10^4
+
+    function getCurrentAuctionPrice(address sellToken, address buyToken, uint auctionIndex) public view
+        returns (uint num, uint den);
+}
+
+
+contract KyberDutchXReserve is KyberReserveInterface, Withdrawable, Utils2 {
+
+    uint public constant BPS = 10000;
+    uint public constant DEFAULT_KYBER_FEE_BPS = 25;
+    uint public feeBps = DEFAULT_KYBER_FEE_BPS;
+    uint public dutchXFeeNum;
+    uint public dutchXFeeDen;
+
+    DutchXExchange public dutchX;
+    address public kyberNetwork;
+    WETH9 public weth;
+
+    mapping(address => bool) listedTokens;
+
+    bool public tradeEnabled;
+
+    /**
+        Constructor
+    */
+    function KyberDutchXReserve(
+        DutchXExchange _dutchX,
+        address _admin,
+        address _kyberNetwork,
+        WETH9 _weth
+    )
+        public
+    {
+        require(address(_dutchX) != 0);
+        require(_admin != 0);
+        require(_kyberNetwork != 0);
+        require(_weth != WETH9(0));
+
+        dutchX = _dutchX;
+        admin = _admin;
+        kyberNetwork = _kyberNetwork;
+        weth = _weth;
+
+        weth.approve(dutchX, 2 ** 255);
+        setDecimals(ETH_TOKEN_ADDRESS);
+        listedTokens[ETH_TOKEN_ADDRESS] = true;
+    }
+
+    function() public payable {
+        // anyone can deposit ether
+    }
+
+    struct AuctionData {
+        uint index;
+        ERC20 srcToken;
+        ERC20 dstToken;
+        uint num;
+        uint den;
+    }
+
+    /**
+        Returns rate = dest quantity / source quantity.
+    */
+    function getConversionRate(
+        ERC20 src,
+        ERC20 dest,
+        uint srcQty,
+        uint blockNumber
+    )
+        public
+        view
+        returns(uint)
+    {
+        blockNumber;
+        if (!tradeEnabled) return 0;
+        if (!listedTokens[src] || !listedTokens[dest]) return 0;
```

Comment:
must check that one of them is ether. and other one is listed token.

Reply:
well
since I know how kyber works, its not necessare.
for bur general case. OK
will do it.